std/typecons.d

```diff
@@ -4865,12 +4865,154 @@ unittest
     assert(rc1 == 5);
     // No more allocation, add just one extra reference count
     auto rc2 = rc1;
+    assert(rc1.refCountedStore.refCount == 2);
     // Reference semantics
     rc2 = 42;
     assert(rc1 == 42);
     // the pair will be freed when rc1 and rc2 go out of scope
 }
 
+// recursively destroy a class instance
+private void _destroy(T)(T obj) if (is(T == class))
+{
+     foreach (C; TypeTuple!(T, BaseClassesTuple!T))
+     {
+         static if (__traits(hasMember, C, "__dtor"))
+             (cast(C)obj).__dtor();
+     }
+     // leak monitor b/c destruction would be impure, will be collected by the GC
+}
+
+/// RefCounted implementation for classes.
+struct RefCounted(T)
+    if (is(T == class))
+{
+    /// $(D RefCounted) storage implementation.
+    struct RefCountedStore
```

Comment:
Is there anything that should be user-facing apart from `refCount`? If not, it could just be moved to `RefCounted` itself. There is a cost to public symbols because of the generic AliasThis, but it already has `refCountedStore` for this, so it would just remove one and add one.